### PR TITLE
[12.0] Add graph of dependencies between jobs

### LINKED ISSUE #129: [Feature] Job dependencies
Some details may be inaccurate as I didn't verify all my assumptions.

# Implementation Draft

First let's state what we would expect from dependencies

## Design assumptions

1. We can delay a job A with a dependency on another job Z: A will start only when Z is in state 'done'
2. We can delay jobs A and B with a dependency on another job Z: both A and B will start only when Z is in state 'done'
3. We can delay a job A with a dependency on other jobs Z and Y: A will start only when Z and Y are in state 'done'
4. We can delay jobs A and B with a dependency on other jobs Z and Y: A will start only when Z and Y are in state 'done'
5. When job Z finishes and A depends on it, A will get the result of Z
6. When job Z and Y finishes and A depends on them, A will get a list with the results of Z and Y
7. When job Z and Y finishes and A and B depends on them, A and B will get a list with the results of Z and Y

Assumptions 1. and 2. are probably the easiest to implement, and we could start with them, but we have to think the datamodel and API to be sure we can do 3. and 4. as well.

## Datamodel and workflow

* Add a new field on queue_job: `dependencies`. This is a JSON field with a dict expressing the dependencies, with 2 keys: `depends_on`, `reverse_depends_on`.  We don't want FKs, jobs are not supposed to be deleted and if they are, we just have to ignore the irrelevant references.
* Add a new field for the result that is passed to the dependencies, JSON serialized (we can pass recordsets, ...). As we currently store the result as a message for the end-user, not sure how we should handle it
* We add a new state `wait_dependency`, these jobs will be ignored by the jobrunner out of the box
* When a job finishes, the job runner looks if the job has a non-empty `reverse_depends_on` list. If it finds dependent jobs, for each job in `reverse_depends_on`, it has to read the `depends_on` list, check if all the jobs in this list are done and in such case, change the state of the dependent job to 'pending'.

Using the denormalization with a `reverse_depends_on` field means that we only never have to run a `SELECT ... WHERE depends_on = ?` to find the dependencies. We have a cost penalty (besides the cost to read `reverse_depends_on`) only when we actually have dependencies.

At first, it does looks not too difficult, the difficulty here lies in the transactions. If a job A depends on Z, when Z is done, it will check the state of the dependencies, Z being the only dependency and being done, we are good. Now if A depends on Z and Y, if Z finishes, it has to check if Y is done as well. Y may or may not be ~dead~ done, but Z being isolated in a transaction will not know the real state of Y. Changing the isolation level would not help as Y can be 'done' then rollbacked. I thought about using postgres 2-phases commit - not sure we can integration it properly in odoo - or opening a new transaction after Z is committed (and take the risk that the second transaction fails, then handle retries, ...).

## API

On the API side, I thought about different options too, I remember I had a preference and some ideas to actually implement them but can't remember ;)
There is much more thoughts to put here, also considering that dependent jobs may receive a result from the jobs they depend on.

### Dependencies API

There is no support of `then` (executed after done or failure), only `done` (executed after `done`).
The "failure" callback is handled synchronously in Celery because you may want to use the traceback from it. This is not drafted here neither. Both of these use cases can be added later, we only have to keep in mind the design to support them (keep this info in the dependencies json field.

Comments in the code explain the additions.

```python
from odoo import _, api, fields, models
from odoo.addons.queue_job.job import job, chain, group


class ProductProduct(models.Model):
    _inherit = 'product.product'

    @api.job
    def generate_thumbnail(self, size):
        return self._gen(size)

    @api.multi
    def button_generate_usual(self):
        # This is the current implementation, which will not change
        self.ensure_one()
        self.with_delay(
            priority=30,
            description=_("generate xxx")
        ).generate_thumbnail((50, 50))

    @api.multi
    def button_generate_simple_with_delayable(self):
        self.ensure_one()
        # Introduction of a delayable object, using a builder pattern
        # allowing to chain jobs or set properties. The delay() method
        # on the delayable object actually stores the delayable objects
        # in the queue_job table
        (
            self.delayable()
            .generate_thumbnail((50, 50))
            .set(priority=30)
            .set(description=_("generate xxx"))
            .delay()
        )

    @api.multi
    def button_generate_chain(self):
        product1 = self.browse(1)
        product2 = self.browse(2)
        product3 = self.browse(3)
        self.ensure_one()
        # the delayable object can be chained
        job1 = (
            product1.delayable().generate_thumbnail((50, 50))
            .set(description=_("generate xxx"))
        )
        job2 = (
            product2.delayable().generate_thumbnail((50, 50))
            .set(description=_("generate xxx")),
        )
        job3 = (
            product3.delayable().generate_thumbnail((50, 50))
            .set(description=_("generate xxx")),
        )
        # the builder pattern allow to set properties at once
        for job_ in (job1, job2, job3):
            # job.set(xxx) is  better than job.priority = 10
            # or than job.priority(10), because it allows to pass
            # a splat dictionary set(**properties) or custom properties
            job.set(priority=30)
        # Or self.delay_chain() as syntactic sugar, calling chain
        # under the hood.
        # .delay() setup the links in the chain and stores the job
        chain(
            job1,
            # job2 is executed after product1
            job2,
            # job3 is executed after product2
            job3
        ).delay()

    @api.multi
    def button_generate_chain_done(self):
        product1 = self.browse(1)
        product2 = self.browse(2)
        product3 = self.browse(3)
        self.ensure_one()
        job1 = product1.delayable().generate_thumbnail((50, 50))
        job2 = product2.delayable().generate_thumbnail((50, 50))
        job3 = product3.delayable().generate_thumbnail((50, 50))
        # similar than the chain
        job1.on_done(job2.on_done(job3)).delay()

    @api.multi
    def button_generate_done_multi(self):
        product1 = self.browse(1)
        product2 = self.browse(2)
        product3 = self.browse(3)
        self.ensure_one()
        job1 = product1.delayable().generate_thumbnail((50, 50))
        job2 = product2.delayable().generate_thumbnail((50, 50))
        job3 = product3.delayable().generate_thumbnail((50, 50))
        # job2 and job3 executed only when job1 is done
        job1.on_done(job2, job3).delay()


class Album(models.Model):
    _name = 'album'

    @api.job
    def generate_page(self, results, color, images_per_page=10):
        return self._gen(results, images_per_page)

    @api.multi
    def button_generate(self):
        self.ensure_one()
        product1 = self.env['product.product'].browse(1)
        product2 = self.env['product.product'].browse(2)
        product3 = self.env['product.product'].browse(3)
        job1 = (
            product1.delayable().generate_thumbnail((50, 50))
            .set(priority=30)
            .set(description=_("generate xxx"))
        )
        job2 = (
            product2.delayable().generate_thumbnail((50, 50))
            .set(priority=30)
            .set(description=_("generate xxx"))
        )
        job3 = (
            product3.delayable().generate_thumbnail((50, 50))
            .set(priority=30)
            .set(description=_("generate xxx"))
        )
        # could have self.delay_group() in base.model as syntactic sugar
        group(job1, job2, job3).on_done(
            # executed only when job1, job2 and job3 are done
            self.delayable().generate_page('#000000', images_per_page=12)
            .set(priority=12)
            .set(description=_("generate page x"))
            # delay makes the links and store in db
        ).delay()

```

### Results API

Still do define

```python
# jobs definitions
from odoo.addons.queue_job import Result

@job
def render_thumbnail(self):
    # could use a class to differentiate the current user message and 
    # the result for the dependent job(s)
    thumbnail = self.env['thumbnail'].do_stuff()
    return Result(message='thumbnail generation is a success',
                          results=thumbnail)
```

### PR DESCRIPTION
**Pull request by guewen**
Issue described in #129.

# Introduction

This PR adds a way to express dependencies between jobs, so called jobs graphs.

The delaying API has been rewritten in a way that allows to express these dependencies. Under the hood, an acyclic direct graph is built and stored in the job records. Jobs which depend on another or several other jobs are in a state `wait_dependencies`, when a job is done, it finds the dependent jobs, which is set to `pending` if all the other dependent jobs are done as well.

# API changes

## Single jobs

The new API is entirely backward compatible. Delaying a single job is done as before with:

```python
records.with_delay(**job_options).my_method(*args, **kwargs)
```

A new concept of delayable is introduced to compose graph, the above is equivalent to:

```python
delayable = records.delayable(**job_options)
delayable.my_method(*args, **kwargs)
delayable.delay()
```

## Chains

A very simple graph, execute a job when the first is done, can be written as:

```python
initial_delayable = records.delayable(**job_options)
initial_delayable.my_method(*args, **kwargs)
initial_delayable.on_done(records.delayable().my_other_method())
initial_delayable.delay()
```

`.delay()` is called only on the root of the graph.

If several jobs are chained, it can be easier to build a chain which are executed in sequence:

```python
from odoo.addons.queue_job.delay import chain

chain(
    records.delayable().my_method(),
    records.delayable().my_second_method(),
   records.delayable().my_third_method(),
).delay()
```

## Groups

Dependencies permit depending on a group of jobs, the third method is executed when the two jobs of the group are done:

```python
from odoo.addons.queue_job.delay import group

group(
    records.delayable().my_method(),
   records.delayable().my_second_method(),
).on_done(records.delayable().my_third_method()).delay()
```

# UI for graphs

Changes for the graphs and some various tweaks:

* The jobs of a graph all share a "Graph UUID". We can search and group jobs per Graph UUID.
* From the form view of a job, we can quickly open all the other jobs of the same graph
* A new widget shows a visualization of the graph in the form view
* The UUID, graph UUID, enqueued time and func_string are now shown only in debug mode

# Improved testing

## Unit tests

A new testing facility is added, which helps a lot to run assertions on enqueued jobs. It is strongly encouraged to write unit tests using this assert. It works on any jobs, this is not specially related to graphs.

Inside the `mock_jobs` context manager, jobs are never stored in database. However, they are kept in memory in the context manager instance, which allow running checks on them.

```python
    # tests
    from odoo.addons.queue_job.tests.common import mock_jobs


    def test_method_to_test(self):
        with mock_jobs() as jobs_tester:
            result = self.env["model"].method_to_test()
            expected_count = 12

            jobs_tester.assert_jobs_count(1, only=self.env["model"].my_job_method)
            jobs_tester.assert_enqueued_job(
                self.env["model"].my_job_method,
                args=("Hi!",),
                kwargs=dict(count=expected_count),
                properties=dict(priority=15)
            )
            self.assertEqual(result, expected_count)
```

Output of failed assertion

Additionally, as the jobs are stored in the context manager instance, we can even choose to execute them in the test or not, depending on what we are testing.

```python
             jobs_tester.perform_enqueued_jobs()
```

## Test jobs

The controller `/queue_job/create_test_job` can now generate a random graph of jobs, using the parameters:

* priority: priority of the created jobs
* max_retries: max_retries of the created jobs
* channel: channel of the created jobs
* description: you guess what
* size: number of jobs delayed, if 1, it will not be a graph
* failure_rate: between 0 and 1, if > 0 jobs will randomly fail according to this rate

# Warnings

As of today, there is no correct handling of identity keys for graphs. We still need to figure out the most correct way to handle an identity key for a graph. However, the identity key still work the same way for single jobs.

Closes #129

### CODE REVIEW

File: queue_job/delay.py [hunk 1]
```diff
@@ -130,9 +131,31 @@ def delay(self):
             vertex._build_job()
         for vertex, neighbour in graph.edges():
             neighbour._generated_job.add_depends({vertex._generated_job})
+
+        # If all the jobs of the graph have another job with the same identity,
+        # we do not create them. Maybe we should check that the found jobs are
+        # part of the same graph, but not sure it's really required...
+        # Also, maybe we want to check only the root jobs.
```
Comment (guewen) [inline queue_job/delay.py:285]: idea: add an uuid field for the whole graph (improve search as well)

**Comment by guewen** [comment]
I pushed new commits. I still have lots of docs to write and not everything is ready yet, but I would welcome feedback by people interested in the feature :) You can look at the doc of the 'base' model methods and at the tests to have an idea of the dependency API.

File: queue_job/job.py [hunk 2]
```diff
@@ -244,13 +196,25 @@ class Job(object):
...
         if not stored:
             raise NoSuchJobError(
                 'Job %s does no longer exist in the storage.' % job_uuid)
         return cls._load_from_db_record(stored)
 
+    @classmethod
+    def load_many(cls, env, job_uuids):
+        """Read jobs in batch from the Database
+
+        Jobs not found are ignored.
```
Comment (guewen) [inline queue_job/job.py:228]: todo: log a warning

File: queue_job/models/queue_job.py [hunk 3]
```diff
@@ -103,6 +107,53 @@ def init(self):
...
+                current = jobs.pop()
+                seen.add(current.id)
+                graph.add_vertex(current.id)
+
+                dependencies = current.dependencies
+                depends_on = dependencies.get('depends_on', [])
+                reverse_depends_on = dependencies.get(
+                    'reverse_depends_on', []
+                )
+                parents = self.search([
```
Comment (guewen) [inline queue_job/models/queue_job.py]: Using a single uuid for the whole graph should improve this a lot.

**Comment by guewen** [comment]
Something I'm really not sure about is how to deal with identity keys. Currently, I'm checking that if a graph contains some jobs with an identity key, if all the jobs in the graph with an identity key have another job in the db with the same key, then the whole graph is discarded. It doesn't seem very intuitive. What would make most sense to me is to have a single identity key for the whole graph. I'll think about how we could implement this.

**Comment by guewen** [comment]
Remaining

* use a single uuid for identifying a graph
* handling of identity keys (not really defined yet)
* handle deprecation of a few things
* implement graph traversal to execute job synchronously when TEST_QUEUE_JOB_NO_DELAY is used
* take care of setting the proper state on requeued children jobs when their parents are not done
* docs
* extend test coverage (I don't remember what is/isn't tested)
* migration to 13.0
* wish: performance analysis, if time allows...
* no "Result" will be handled yet, it can be added later

**Comment by etobella** [comment]
@guewen Do you need some help with this?

**Comment by guewen** [comment]
@etobella The only thing yet to do I think is about handling the identity key for a whole graph. Well, there is an implementation now which is that if all jobs have an identity key and all these identity keys already exist, we skip the graph. I don't think this is how it should be done, rather have a single identity key for the whole graph.

Anyway, I think this could be handled later, the "how" could become more clear at usage.

This PR is in a safe enough state to be tested if you'd like :) I should open my PR soon, I'd like to improve the description first 😆

**Comment by simahawk** [comment]
Beautiful work, as usual :)
Question: any chance to add this feature to a separated module?
It's quite big as change and it will be easier to maintain (and to maintain the core) if it stays separated.

**Comment by guewen** [comment]
> Question: any chance to add this feature to a separated module?

This is not something "on top" of the existing API, but an improvement of the current API to support graphs. Enqueuing a single job is now (aside a few details) enqueueing a graph of 1 job, the mechanism is generic for supporting both (`with_delay` goes through the graph machinery - syntactic sugar when you want only 1 job -, but the performance is the same as today when you have 1 job).  In other words: this is the core now ;)

File: queue_job/views/queue_job_views.xml [hunk 4]
```diff
@@ -26,16 +26,27 @@
...
+                        statusbar_visible="wait_dependencies,pending,enqueued,started,done"
                         statusbar_colors='{"failed":"red","done":"green"}'/>
                 </header>
                 <sheet>
+                    <div class="oe_button_box" name="button_box">
+                        <button
+                            name="open_graph_jobs"
+                            type="object"
+                            class="oe_stat_button"
+                            icon="fa-pencil-square-o"
```
Comment (guewen) [inline queue_job/views/queue_job_views.xml]: todo: find a better icon

Comment (etobella) [reply queue_job/views/queue_job_views.xml]: On Fontawesome 5 I would propose https://fontawesome.com/v5.15/icons/project-diagram?style=solid

With Fontawesome 4.7 it is hard to find something, as there is not many options :disappointed:

Comment (gregory-moka) [reply queue_job/views/queue_job_views.xml]: Maybe : https://fontawesome.com/v4.7/icon/sitemap

Comment (guewen) [reply queue_job/views/queue_job_views.xml]: The sitemap is great, thanks!

**Comment by guewen** [comment]
Anyone interested in this?
The PR is ready ;)

**Review by sbidoul — APPROVED** [review]
Fantastic work @guewen ! This is going to be extremely useful.

Had you considered 'then' instead of 'done' ?

A word in the docs about what happens when a job of graph fails might be helpful.

**Comment by guewen** [comment]
Thanks @sbidoul! 
It has been a work on the long run as I did the first commit on the 2nd of July 2019 and the last commits on the 2nd of November 2021 :)

> Had you considered 'then' instead of 'done' ?

I remember I did! But I don't remember why I preferred `done`. But if I think about it again, I would say that in the jquery promises (where I probably took the inspiration?), `then` will be called for both success and failures, and `done` only for success, and I wanted to emphasize on this. Nowadays, jquery is less and less used and I don't think promises have a `done` function.

Do you think it would be better with `then`? It would be an easy change that I'll be happy to do if it makes the API better.

> A word in the docs about what happens when a job of graph fails might be helpful.

Yep, I'll do that.

BTW, I just pushed the 14.0 version here https://github.com/OCA/queue/pull/403

And while I'm here, @sbidoul, outstanding work on runboat :green_heart:

**Comment by sbidoul** [comment]
> Do you think it would be better with then? It would be an easy change that I'll be happy to do if it makes the API better.

My first impression was that `done()` could create some confusion with `set_done()`.
And Indeed js promises have `then()`, `catch()` and `finally()`. Although js `then` accepts a second argument for then rejected case. Not sure... up to you :)

**Comment by simahawk** [comment]
Could be done more verbose like `on_done` and `on_fail`.
Regarding using `then` I'm not sure it will add clarity :)
But yeah... not a game changer in this wonderful work :)

@guewen do you prefer to merge 12.0 first?

**Comment by guewen** [comment]
`on_done()` is maybe a better choice, it is less ambiguous than `then()` (clearer on the fact it happens on `done()`) and `done()` (less confusion with `set_done()`). I'll do the renaming soon, unless you have other suggestions.

I was thinking about the naming of `mock_jobs()` too, and renamed it to `trap_jobs()`, because I think the latter describes better what *it does*, vs *how it does it* for the former (which is a hidden technical detail).

**Comment by guewen** [comment]
> @guewen do you prefer to merge 12.0 first?

I suppose I'll do the last renaming changes here, then cherry-pick them on 14.0, so yes it'd make sense.

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
seems it failed... 
BTW are you planning to fwd port?

/ocabot merge major

**Comment by sbidoul** [comment]
Apparently travis did not report back ? I take the opportunity to test the freshly merged OCA repo tempalte for 12.0 in #426, and move to GitHub actions.

**Comment by sbidoul** [comment]
Ok, done. @guewen can you rebase ?

**Comment by guewen** [comment]
@simahawk version for 14.0 is there: https://github.com/OCA/queue/pull/403
I didn't work on any other version yet.

**Pull request merged**